matchpygen/example_yield.cpp

```diff
@@ -0,0 +1,253 @@
+#include <iostream>
+#include <tuple>
+#include <map>
+#include <string>
+#include <deque>
+#include <functional>
+
+#include <symengine/basic.h>
+#include <symengine/pow.h>
+
+using namespace std;
+using namespace SymEngine;
+
+typedef map<string, RCP<const Basic>> Substitution2;
+
+int try_add_variable(Substitution2 &subst, string variable_name,
+                     RCP<const Basic> &replacement)
+{
+    if (subst.find(variable_name) == subst.end()) {
+        subst[variable_name] = replacement;
+        // this[variable_name] = replacement; //.copy() if
+        // isinstance(replacement, Multiset) else replacement
+    } else {
+        // Node &existing_value = *(this->find(variable_name));
+        /*
+         if (isinstance(existing_value, tuple)) {
+         if (isinstance(replacement, Multiset)) {
+         if (Multiset(existing_value) != replacement){
+         return -1;
+         }
+         } else if (replacement != existing_value) {
+         return -1;
+         }
+         else if (isinstance(existing_value, Multiset)) {
+         if (!isinstance(replacement, (tuple, list, Multiset))) {
+         return -1;
+         }
+         auto &compare_value = Multiset(replacement);
+         if (existing_value == compare_value){
+         if (! isinstance(replacement, Multiset)) {
+         this->insert(make_pair(variable_name, replacement));
+         } else {
+         return -1;
+         }
+         } else if (replacement != existing_value){
+         return -1;
+         }
+         }
+         }
+         */
+    }
+    return 0;
+}
+
+typedef deque<RCP<const Basic>> Deque;
+
+Deque get_deque(RCP<const Basic> expr)
+{
+    Deque d;
+    for (RCP<const Basic> i : expr->get_args()) {
+        d.push_back(i);
+    }
+    return d;
+}
+
+class match_root
+{
+public:
+    typedef tuple<int, Substitution2> YieldType;
+    void (match_root::*current)();
```

Comment:
I think you should use the `std::function` for this to avoid handling raw pointers.

Reply:
I guess you're right. I'll give it a try later.

Reply:
So, using `std::function` would introduce a double overhead. You need to assign a method like:
```c++
current = std::bind(&match_root::next_function, this);
```
or 
```c++
current = [=](){ this->next_function(); };
```
I would stick with the function pointers.